### PR TITLE
--enable-experiments for `pub run`

### PR DESCRIPTION
**Pull request by sigurdm**
Fixes: https://github.com/dart-lang/pub/issues/2492, https://github.com/dart-lang/pub/issues/2360, https://github.com/dart-lang/sdk/issues/41355 .

Run as a subprocess if there are vm-args, but keep using isolates in the default case due to better signal handling.

Also --vm-arg that passes arguments to the VM for `pub run`.
`pub run` now runs programs as sub-processes.

We do not run from snapshot if running with non-default vm args.

Snapshots are now saved with the sdk-version embedded in the name, this
avoids the awkward detection when the vm fails running a bad snapshot.

Contains quite some refactorings. Especially streamlining of `pub global
run`.

### CODE REVIEW

File: lib/src/global_packages.dart [hunk 1]
```diff
@@ -605,7 +527,9 @@ class GlobalPackages {
       var script = package.pubspec.executables[executable];
 
       var previousPackage = _createBinStub(package, executable, script,
-          overwrite: overwriteBinStubs, snapshot: snapshots[script]);
+          overwrite: overwriteBinStubs,
+          snapshot: entrypoint
+              .snapshotPathOfExecutable(Executable(package.name, executable)));
```
Comment (kevmoo) [inline lib/src/global_packages.dart:532]: Adding a trailing comma to the `_creatBinStub` args will make this format nicer 😄

File: lib/src/entrypoint.dart [hunk 2]
```diff
@@ -274,59 +288,98 @@ class Entrypoint {
...
+  /// The location of the snapshot of the dart program at [path] in [package]
+  /// will be stored here.
+  ///
+  /// We use the sdk version to make sure we don't run snapshots from a
+  /// different sdk.
+  ///
+  /// [path] must be relative.
+  String snapshotPathOfExecutable(Executable executable) {
+    assert(p.isRelative(executable.relativePath));
+    final versionSuffix = Uri.encodeComponent(Platform.version);
```
Comment (jonasfj) [inline lib/src/entrypoint.dart:369]: See `hosted.dart` where we have:
```dart
  /// The system cache directory for the hosted source contains subdirectories
  /// for each separate repository URL that's used on the system.
  ///
  /// Each of these subdirectories then contains a subdirectory for each
  /// package downloaded from that site.
  @override
  String getDirectory(PackageId id) {
    var parsed = source._parseDescription(id.description);
    var dir = _urlToDirectory(parsed.last);
    return p.join(systemCacheRoot, dir, '${parsed.first}-${id.version}');
  }
```

This is probably unnecessary...

File: test/snapshot_test.dart [hunk 3]
```diff
@@ -222,11 +224,11 @@ void main() {
...
         await d.appDir({'foo': '5.6.7'}).create();
 
-        await pubGet(
-            args: ['--precompile'], output: contains('Precompiled foo:hello.'));
+        await pubGet(args: ['--no-precompile']);
 
         await d.dir(p.join(appPath, '.dart_tool', 'pub', 'bin'), [
-          d.dir('foo', [d.outOfDateSnapshot('hello.dart.snapshot.dart2')])
+          d.dir('foo',
+              [d.outOfDateSnapshot('hello.dart.snapshot.2.0.0-dev.58.0')])
```
Comment (jonasfj) [inline test/snapshot_test.dart:231]: No `$versionSuffix`

File: lib/src/package.dart [hunk 4]
```diff
@@ -34,6 +34,8 @@ class Package {
   /// The path to the directory containing the package.
   final String dir;
 
+  bool get isInMemory => dir == null;
```
Comment (jonasfj) [inline lib/src/package.dart:37]: This doesn't have to be public does it? Can we also give a single line doc-comment.

File: lib/src/package_graph.dart [hunk 5]
```diff
@@ -39,7 +39,8 @@ class PackageGraph {
   /// the packages' pubspecs are already fully-parsed.
   factory PackageGraph.fromSolveResult(
       Entrypoint entrypoint, SolveResult result) {
-    var packages = Map<String, Package>.fromIterable(result.packages,
+    var packages = Map<String, Package>.fromIterable(
+        result.packages.where((id) => id.source != null),
```
Comment (jonasfj) [inline lib/src/package_graph.dart:43]: Why is this necessary? Are we abusing notation somehow?

What does `id.source == null` imply?

Comment (jonasfj) [reply lib/src/package_graph.dart:43]: So `root` is filtered out? why?

And why do we retain `if (id.name == entrypoint.root.name) return entrypoint.root;`.

File: lib/src/executable.dart [hunk 6]
```diff
@@ -57,130 +56,152 @@ Future<int> runExecutable(Entrypoint entrypoint, String package,
...
+/// An executable in a package
+class Executable {
+  String package;
+  // The relative path to the executable inside the root of [package].
+  String relativePath;
+  Executable(this.package, String program)
+      : relativePath = adaptProgramToPath(program);
 
-  return exitCode;
+  static String adaptProgramToPath(String program) {
```
Comment (jonasfj) [inline lib/src/executable.dart:196]: This can be private if we call it from the constructor...

but I wanted to avoid entangling this logic in `pub run --dart-dev`. I agree that `pub run` needs to retain this for compatibility. Can we instead of calling it from `Executable` constructor, call it before we pass to `Executable` constructor and then make it public...

Or make it private and give `Executable` two constructors.

File: lib/src/executable.dart [hunk 7]
```diff
@@ -57,130 +56,152 @@ Future<int> runExecutable(Entrypoint entrypoint, String package,
...
+    // Since we don't access the package graph, this doesn't happen
+    // automatically.
+    entrypoint.assertUpToDate();
 
+    if (!fileExists(snapshotPath)) {
+      await recompile(executable);
+    }
+    if (!fileExists(snapshotPath)) {
+      // TODO XXX compilation failed, run from source??
+    }
```
Comment (jonasfj) [inline lib/src/executable.dart:90]: I think `dart.snapshot` will throw `ApplicationException` if the compilation fails.. SP there is no need to handle that here.

File: lib/src/command/global_run.dart [hunk 8]
```diff
@@ -57,8 +61,18 @@ class GlobalRunCommand extends PubCommand {
...
+    final vmArgs = <String>[
+      if (experiments.isNotEmpty) "--enable_experiment=${experiments.join(',')}"
+    ];
+    final globalEntrypoint = await globals.find(package);
+    final exitCode = await runExecutable(
+        globalEntrypoint, Executable(package, executable), args,
+        vmArgs: vmArgs,
+        enableAsserts: argResults['enable-asserts'] || argResults['checked'],
+        recompile: (executable) async {
+      await globalEntrypoint.precompileExecutable(executable);
```
Comment (jonasfj) [inline lib/src/command/global_run.dart:74]: `recompile: globalEntrypoint.precompileExecutable` wouldn't work? I don't think you need to wrap it.

File: lib/src/entrypoint.dart [hunk 9]
```diff
@@ -122,31 +123,47 @@ class Entrypoint {
 
   PackageGraph _packageGraph;
 
+  /// Where the lock file and package configurations are to be found.
+  ///
+  /// Global packages (except those from path source)
+  /// store these in the global cache.
+  String get _configPath => root.dir == null || isCached
```
Comment (jonasfj) [inline lib/src/entrypoint.dart:130]: Should we call this `_configRoot`?

File: lib/src/entrypoint.dart [hunk 10]
```diff
@@ -122,31 +123,47 @@ class Entrypoint {
...
-    var newPath = root.path('.dart_tool/pub');
-    var oldPath = root.path('.pub');
-    if (!dirExists(newPath) && dirExists(oldPath)) return oldPath;
-    return newPath;
+    if (isGlobal) {
+      return p.join(
+        cache.rootDir,
+        'global_packages',
+        root.name,
+      );
```
Comment (jonasfj) [inline lib/src/entrypoint.dart:160]: Why is this not using `_configPath`.

And why is this different for global packages installed with `pub global activate --source path`

File: lib/src/entrypoint.dart (hunk 2 above)
Comment (jonasfj) [reply lib/src/entrypoint.dart:369]: After chatting we agree to use the version number :D

File: lib/src/package_graph.dart (hunk 5 above)
Comment (jonasfj) [reply lib/src/package_graph.dart:43]: Okay, let's move this special filtering hack into `_installInCache` in `global_packages.dart`.

File: lib/src/executable.dart (hunk 6 above)
Comment (jonasfj) [reply lib/src/executable.dart:196]: Let's make two constructors...

File: lib/src/entrypoint.dart (hunk 10 above)
Comment (jonasfj) [reply lib/src/entrypoint.dart:160]: Documentation comment about why this is different would be nice.

File: lib/src/command/global_run.dart [hunk 11]
```diff
@@ -57,8 +61,16 @@ class GlobalRunCommand extends PubCommand {
       log.warning('The --mode flag is deprecated and has no effect.');
     }
 
-    var exitCode = await globals.runExecutable(package, executable, args,
-        enableAsserts: argResults['enable-asserts'] || argResults['checked']);
+    final experiments = argResults['enable-experiment'] as List;
+    final vmArgs = <String>[
+      if (experiments.isNotEmpty) "--enable_experiment=${experiments.join(',')}"
```
Comment (kevmoo) [inline lib/src/command/global_run.dart:66]: Silly nit: use `enable-experiment` for the SDK. It supports both `_` and `-` for separators, but we should be consistant!

File: lib/src/command/run.dart [hunk 12]
```diff
@@ -67,28 +70,16 @@ class RunCommand extends PubCommand {
       log.warning('The --mode flag is deprecated and has no effect.');
     }
 
-    // The user may pass in an executable without an extension, but the file
-    // to actually execute will always have one.
-    if (p.extension(executable) != '.dart') executable += '.dart';
+    final experiments = argResults['enable-experiment'] as List;
+    final vmArgs = <String>[
+      if (experiments.isNotEmpty) "--enable_experiment=${experiments.join(',')}"
```
Comment (kevmoo) [inline lib/src/command/run.dart:75]: ditto

File: lib/src/command/run.dart [hunk 13]
```diff
@@ -126,41 +118,15 @@ class RunCommand extends PubCommand {
...
-
-    // Don't ever compile snapshots for mutable packages, since their code may
-    // change later on. Don't check if this the case if a snapshot already
-    // exists.
-    var useSnapshot = snapshotExists ||
-        (package != entrypoint.root.name &&
-            !entrypoint.packageGraph.isPackageMutable(package));
+    final experiments = argResults['enable-experiment'] as List;
+    final vmArgs = <String>[
+      if (experiments.isNotEmpty) "--enable_experiment=${experiments.join(',')}"
```
Comment (kevmoo) [inline lib/src/command/run.dart:123]: ditto

Comment (kevmoo) [reply lib/src/command/run.dart:123]: Seeing this logic three times, wondering if it's worth a utility 🤔

File: lib/src/command/global_run.dart [hunk 14]
```diff
@@ -27,6 +28,9 @@ class GlobalRunCommand extends PubCommand {
   GlobalRunCommand() {
     argParser.addFlag('enable-asserts', help: 'Enable assert statements.');
     argParser.addFlag('checked', abbr: 'c', hide: true);
+    argParser.addMultiOption('enable-experiment',
+        help: 'Runs the executable in a VM with the given experiments enabled.',
```
Comment (kevmoo) [inline lib/src/command/global_run.dart:32]: 1) Wondering if we should include details saying that the target executable won't be precompiled, so it'll start up more slowly
2) Is there somewhere  we can point folks to get more info on experiments?

https://github.com/dart-lang/sdk/blob/master/docs/process/experimental-flags.md is not super helpful

CC @mit-mit

Comment (mit-mit) [reply lib/src/command/global_run.dart:32]: Og, maybe we should add a short page on dart.dev about experiment flags, WDYT @kwalrath ?

Comment (kevmoo) [reply lib/src/command/global_run.dart:32]: Well, I think linking to the github repo is fine – maybe we could have a redirect like flutter.dev/go/... right?

Comment (kwalrath) [reply lib/src/command/global_run.dart:32]: Sure, we could have a page on dart.dev. Where would it go? What should it cover?

File: lib/src/command/run.dart (hunk 13 above)
Comment (sigurdm) [reply lib/src/command/run.dart:123]: I made a function!

File: lib/src/command/run.dart [hunk 15]
```diff
@@ -101,7 +90,8 @@ class RunCommand extends PubCommand {
   ///
   /// Runs `bin/<command>.dart` from package `<package>`. If `<package>` is not
   /// mutable (local root package or path-dependency) a source snapshot will be
-  /// cached in `.dart_tool/pub/bin/<package>/<command>.dart.snapshot.dart2`.
+  /// cached in
+  /// `.dart_tool/pub/bin/<package>/<command>.dart.snapshot.$versionSuffix`.
```
Comment (jonasfj) [inline lib/src/command/run.dart:94]: Can we say `<sdkVersion>`?

Comment (sigurdm) [reply lib/src/command/run.dart]: Done

File: lib/src/command/run.dart [hunk 16]
```diff
@@ -29,9 +29,8 @@ class RunCommand extends PubCommand {
     argParser.addFlag('enable-asserts', help: 'Enable assert statements.');
     argParser.addFlag('checked', abbr: 'c', hide: true);
     argParser.addMultiOption('enable-experiment',
-        help:
-            'Runs the executable in a VM with the given experiments enabled.\n'
-            '(Enabling experiments will disable source snapshot caching)',
+        help: 'Runs the executable in a VM with the given experiments enabled. '
+            '(Enabling experiments will disable source snapshot caching.)',
```
Comment (mit-mit) [inline lib/src/command/run.dart:33]: Can we say something simpler like "(Note: pub commands will run slower when using experiments)"?

Comment (sigurdm) [reply lib/src/command/run.dart]: That's very vague. Then I'll rather say nothing.

Comment (mit-mit) [reply lib/src/command/run.dart]: Then say something more specific, but describe the effect (what the user will experience), not the underlying cause of that effect.

File: lib/src/command/global_run.dart (hunk 14 above)
Comment (mit-mit) [reply lib/src/command/global_run.dart]: @kwalrath i sent some notes in email

**Comment by sigurdm** [comment]
Not sure why travis is not checking the last commit. It should be good to go.

**Comment by kevmoo** [comment]
You could create a branch w/ this PR contents and enable travis for in in the .travis.yml file – just to sanity check!

**Comment by kevmoo** [comment]
Just pushed `travis` branch and it kicked things for you! https://travis-ci.org/github/dart-lang/pub/builds/693915946

File: lib/src/command/run.dart (hunk 16 above)
Comment (jonasfj) [reply lib/src/command/run.dart]: Maybe -> `(Experiments will skip snapshot cache, this might cause slower startup)`

File: lib/src/command/global_run.dart [hunk 17]
```diff
@@ -27,6 +28,10 @@ class GlobalRunCommand extends PubCommand {
   GlobalRunCommand() {
     argParser.addFlag('enable-asserts', help: 'Enable assert statements.');
     argParser.addFlag('checked', abbr: 'c', hide: true);
+    argParser.addMultiOption('enable-experiment',
+        help: 'Runs the executable in a VM with the given experiments enabled. '
+            '(Will disable snapshotting, resulting in slower startup)',
```
Comment (jonasfj) [inline lib/src/command/global_run.dart:33]: Sounds good to me too :D

**Comment by kevmoo** [comment]
Can we land this? Nowish? @sigurdm – still around?

File: lib/src/rate_limited_scheduler.dart [hunk 18]
```diff
@@ -59,7 +60,7 @@ class RateLimitedScheduler<J, V> {
   final Set<J> _started = {};
 
   RateLimitedScheduler(Future<V> Function(J) runJob,
-      {maxConcurrentOperations = 10})
+      {@required int maxConcurrentOperations})
```
[the remark this answers is not in the document]
Comment (jonasfj) [reply lib/src/rate_limited_scheduler.dart:63]: hehe, I guess @sigurdm opted to remove the default value in the function definition, instead of a removing the value at the callsite.. so merging the two branches I had to pick one :D

I'm not sure it matters a lot. But I suppose it's nice to explicit in this case.
